scripts/typescript/src/bcdb/bcdb.ts

```diff
@@ -0,0 +1,352 @@
+      if (index !== null) log_buffer.write(`ERR: line ${index}: ${message}\n`);
+      else log_buffer.write(`ERR: ${message}\n`);   */
+    }
+
+    // Format messages
+    function log(message: string) {    /*  
+      log_buffer.write(`MSG: ${message}\n`);*/
+    }
+
+    //-- Load desired sheets
+
+    log(`=-=-=-=-=-=-=-=-=-= RCBC XLSX to JSON =-=-=-=-=-=-=-=-=`);
+
+    const workBook = XLSX.readFile(options.input_file_path, {
+      sheets: options.sheets,
+      sheetRows: options.limitSheetRows,
+      cellHTML: false,
+      cellFormula: false,
+      dense: true, // really usefull to remove a lot of useless data
+    });
+    log(`OK`);
+
+    //-- Load the sheets
+
+    const project = {
+      name: "BCDB",
+      model_version: "1.0.0",
+      node: ["Project"],
+      collection: new Array<ICollection>(),
+      material: new Array<IMaterial>(),
+    } satisfies IProject;
+
+    // Store references as: doi => IReference
+    const references = new Map<string, IReference>();
+    // Store blocks as: bigsmiles => {...}
+    const blocks = new Map<string, { index: number; polymer: string; bigsmiles: string; name: string }>();
+
+    //-- Loads polymer/name/bigsmiles table
+    {
+      const sheetName = options.sheets[0];
+      process.stdout.write(`Loading sheet ${sheetName} ...\n`);
+      const rows: { [key: string]: string }[] = XLSX.utils.sheet_to_json(workBook.Sheets[sheetName], {
+        header: "A",
+      });
+
+      log(`-- Found ${rows.length} row(s)`);
+
+      rows
+        .slice(1) // Skip header row.
+        .forEach((row, index) => {
+          const polymer = row["A"];
+          const name = row["B"];
+          const bigsmiles = row["C"];
+          const entry = { index, polymer, name, bigsmiles };
+          blocks.set(bigsmiles, entry);
+        });
+      log(`Found ${JSON.stringify([...blocks.values()], null, " ")} (${blocks.size} rows)`);
+      log(`Storing blocks meta information ... DONE`);
+    }
+
+    {
+      const sheetName = options.sheets[1];
+      process.stdout.write(`Loading sheet ${sheetName} ...\n`);
+      const rows: { [key: string]: string }[] = XLSX.utils.sheet_to_json(workBook.Sheets[sheetName], {
+        header: "A",
+      });
+      log(`-- Found ${rows.length} row(s)`);
+      log(`-- Converting polymer rows to JSON ...`);
+
+      rows
+        .slice(2) // Skip 2 header rows
+        .forEach((row, index, arr) => {
+          index % 100 == 0 && process.stdout.write(`-- Loading row ${index}/${arr.length - 1}\n`);
+
+          //-- reference
+          //   to store DOI and ORCIDs
+          const doi = row[Column.DOI];
+          const author = row[Column.ORCID];
+          let reference: IReference | undefined = references.get(doi);
+          if (!reference) {
+            reference = {
+              node: ["Reference"],
+              doi: doi,
+              author: [author], // TODO: parse array
+              type: "", // TODO: pick a reference-type key from vocab
```

Comment:
I am not sure which one to pick, because it is an algorithm but I am citing a paper (doi with ORCID).

Reply:
journal_article